### PR TITLE
Fix various issues when projection is enabled by default

### LINKED ISSUE #34273: Projections and engine Merge
```sql
create table pp(A Int64)Engine=MergeTree order by tuple() as select * from numbers(1e7);

select count() from merge(currentDatabase(), '^pp$') ;
┌──count()─┐
│ 10000000 │
└──────────┘
1 rows in set. Elapsed: 0.015 sec. Processed 10.00 million rows, 80.00 MB (647.12 million rows/s., 5.18 GB/s.)


select count() from merge(currentDatabase(), '^pp$') settings allow_experimental_projection_optimization=1;

Received exception from server (version 22.1.3):
Code: 43. DB::Exception: Received from localhost:9000. DB::Exception: Illegal type
 AggregateFunction(count) of argument of function toUInt64: while converting source column `count()` 
to destination column `count()`. (ILLEGAL_TYPE_OF_ARGUMENT)
```

### PR DESCRIPTION
**Pull request by amosbird**
Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix various issues when projection is enabled by default. Each issue is described in separate commit. This is for https://github.com/ClickHouse/ClickHouse/issues/33678 . This fixes https://github.com/ClickHouse/ClickHouse/issues/34273

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by amosbird** [comment]
Convert to draft for now because the last commit is to enable projection by default, which is used to run CI tests. When it's clean, the last commit will be reverted. We will have dedicated PR to enable projection by default.

**Comment by amosbird** [comment]
It's lucky that we have minmax_count projection enabled for every MergeTree* tables, which helps CI find many corner cases.

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 1]
```diff
@@ -4833,18 +4895,32 @@ std::optional<ProjectionCandidate> MergeTreeData::getQueryProcessingStageWithAgg
...
-                = analysis_result.before_aggregation ? analysis_result.before_aggregation : analysis_result.before_order_by;
-            NameSet required_columns;
-            for (const auto & column : actions->getRequiredColumns())
-                required_columns.insert(column.name);
+            if (analysis_result.before_aggregation && analysis_result.optimize_aggregation_in_order)
+            {
+                for (const auto & key : keys)
+                {
+                    auto actions_dag = analysis_result.before_aggregation->clone();
+                    actions_dag->foldActionsByProjection({key}, sample_block_for_keys);
```
Comment (KochetovNicolai) [inline src/Storages/MergeTree/MergeTreeData.cpp:4905]: May I ask to add a comment to `foldActionsByProjection`, `reorderAggregationKeysForProjection` and `addAggregatesViaProjection` methods of `ActionDAG`?

**Review by KochetovNicolai — COMMENTED** [review]
LGTM,
but, please, add more comments whenever it's possible

File: src/Storages/MergeTree/MergeTreeData.cpp (hunk 1 above)
Comment (amosbird) [reply src/Storages/MergeTree/MergeTreeData.cpp]: Done

**Comment by amosbird** [comment]
Perfect green CI. @KochetovNicolai please take a look if it's ready to merge. I'll revert `Enable projection by default`.

**Review by KochetovNicolai — APPROVED** [review]

**Pull request merged**